Move Fuse to timeseries module

daniel-zullo-frequenz opened the pull request:
The Grid class is the only class that uses `Fuse` as it is used to represent the fuse of a grid. It is not a microgrid-specific class. Therefore, it is moved to the timeseries module.

Fixes #771

daniel-zullo-frequenz committed:
Move Fuse to timeseries module

The Grid class is the only class that uses `Fuse` as it is used
to represent the fuse of a grid. It is not a microgrid-specific class.
Therefore, it is moved to the timeseries module.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Break circular dependency between Current and Fuse

The circular dependency between `Current` and `Fuse` is causing
issues when `component` is imported.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

llucax commented on src/frequenz/sdk/microgrid/component/_component.py:
```diff
@@ -2,16 +2,19 @@
-from ...timeseries import Current
-from ..fuse import Fuse
+if TYPE_CHECKING:
+    # Break circular import by enclosing these type hints in a `TYPE_CHECKING` block.
```
Super silly, but the comment seems a bit too redundant.

```suggestion
    # Break circular import
```

daniel-zullo-frequenz replied:
Added a new commit to amend this comment and similar ones

llucax reviewed (commented):
Is it possible to add a test case to make sure the circular import is not introduced again?

daniel-zullo-frequenz committed:
Test that all modules can be imported

Add an integration test that imports all modules to ensure
that all modules can be imported and that there are no
circular imports.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Amend redundant comment to break circular import

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
> Is it possible to add a test case to make sure the circular import is not introduced again?

Added a test that import all modules, this was enough to reproduce the circular import error that @Marenz observed. It's not a unit test so I added as an integration test

llucax commented on tests/utils/test_integration.py:
This looks like it could replace the very silly `test/test_sdk.py`. I would also call it something more explicit, like `tests/test_import_all.py`, not convinced it we should add the `_integration` suffix too or not :thinking:

daniel-zullo-frequenz replied:
Right, it should replace `test/test_sdk.py`. I think it's fine to remove `_integration` suffix as we are tagging the tests and ideally we shouldn't mix unit and integration in the same file even though there is no restriction to do so in pytest

daniel-zullo-frequenz replied:
Updated as suggested

llucax replied:
> Right, it should replace `test/test_sdk.py`. I think it's fine to remove `_integration` suffix as we are tagging the tests and ideally we shouldn't mix unit and integration in the same file even though there is no restriction to do so in pytest 

The advantage of separating unit and integration tests is that integration tests usually need to import a lot more stuff (and this test case in particular is the extreme case of that 😂). If there is an issue in the extra imports, then the unit tests won't run either, which is confusing.

I'd rather keep them separated to be honest.

daniel-zullo-frequenz replied:
Renamed the test file from `test_import_all.py` to `test_import_all_integration.py`

llucax commented on tests/utils/test_integration.py:
```diff
@@ -0,0 +1,26 @@
+import os
+
+
+def test_all_modules_importable() -> None:
```
It looks like you forgot the `@pytest.mark.integration` here.

daniel-zullo-frequenz replied:
yes, I did. Good catch!

daniel-zullo-frequenz replied:
Fixed

daniel-zullo-frequenz committed:
Remove tests no longer needed

The `tests/test_sdk.py` tests are no longer needed,
as they are fully covered in the recently added integration
tests to check that all modules can be imported.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

Marenz commented:
I just tested my actor with this branch and the circular imports problem is now gone.

daniel-zullo-frequenz commented:
Renamed the test file from `test_import_all.py` to `test_import_all_integration.py`
